File: week-2/happy-number/happy-number.js
/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function(n) {
    const hashtable = {}
    while(n !== 1) {
        if (hashtable.hasOwnProperty(n)) return false;
        
        hashtable[n] = true;
        
        let sum_digits = 0;
        while(n) {
            sum_digits += (n%10)**2;
            n = Math.floor(n/10);
        }
        n = sum_digits;
    }
    
    return true;
};

File: week-2/happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        hashtable = {}
        while n != 1:
            if n in hashtable:
                return False
            hashtable[n] = True
            
            sum_digits = 0
            while n:
                sum_digits += (n%10)**2
                n //= 10
            n = sum_digits
        
        return True
